forest.hs

```diff
@@ -1,40 +1,41 @@
-{-# LANGUAGE BangPatterns #-}
-import Data.List
+module Main where
+
+import qualified Data.Set as S
 import System.Environment
 
 data Forest = Forest
   { goats :: !Int
   , wolves :: !Int
   , lions :: !Int
```

Comment:
Those are strictness annotations for the records, they basically make all pattern matching use bang patterns without needing to write them out? That's as far as I understand at least. See http://book.realworldhaskell.org/read/profiling-and-optimization.html about halfway down the page titled "Bang Patterns" then the section below